src/modules/rest/extjs/src/main/java/it/geosolutions/geostore/services/rest/impl/RESTExtJsServiceImpl.java

```diff
@@ -705,60 +756,62 @@ private void readSecurity() {
-                } else {
-                    // get security rules for groups
-                    List<String> groups = new ArrayList<String>();
-                    for (UserGroup g : authUser.getGroups()) {
-                        groups.add(g.getGroupName());
-                    }
-                    for (SecurityRule rule :
-                            resourceService.getGroupSecurityRule(groups, r.getId())) {
-                        // GUEST users can not access to the delete and edit(resource,data blob is
-                        // editable) services
-                        // so only authenticated users with
-                        if (rule.isCanWrite() && !authUser.getRole().equals(Role.GUEST)) {
-                            canEdit = true;
-                            canDelete = true;
-                            break;
-                        }
-                    }
-                    // get security rules for user
-                    for (SecurityRule rule :
-                            resourceService.getUserSecurityRule(authUser.getName(), r.getId())) {
-                        User owner = rule.getUser();
-                        UserGroup userGroup = rule.getGroup();
-                        if (owner != null) {
-                            if (owner.getId().equals(authUser.getId())) {
-                                if (rule.isCanWrite()) {
-                                    canEdit = true;
-                                    canDelete = true;
-                                    break;
-                                }
-                            }
-                        } else if (userGroup != null) {
-                            if (authUser.getGroups() != null
-                                    && authUser.getGroups().stream()
-                                            .noneMatch(
-                                                    ug ->
-                                                            ug.getGroupName()
-                                                                    .equals(
-                                                                            userGroup
-                                                                                    .getGroupName()))) {
-                                if (rule.isCanWrite()) {
-                                    canEdit = true;
-                                    canDelete = true;
-                                    break;
-                                }
-                            }
-                        }
-                    }
                 }
             }
         }
 
+        private boolean canUserGroupsEditResource() {
+            List<String> groupsNames =
+                    authUser.getGroups().stream()
+                            .map(UserGroup::getGroupName)
+                            .collect(Collectors.toList());
+            return resourceService.getGroupSecurityRule(groupsNames, r.getId()).stream()
```

Comment:
This is a direct DB call for each resource. If getExtResourcesList() is returning hundreds or thousands of resources, each resource may lead to repeated queries.
Evaluate how frequently you call these “per resource” methods. If the code is triggered in a big listing, you may want a bulk approach to fetch all SecurityRules relevant to that batch of resources.

Reply:
Fixed. Thanks!